### PR TITLE
re-did the crawler

### PR DESCRIPTION
**Pull request by c-rotte**
**DRAFT**
- Problem: The current crawler does not capture all modules due to a (imo) design flaw in the curricula
- Things TODO:
- [x] re-design the database structure
- [x] integrate the database structure in the crawler
- [x] change the api endpoints
- [x] api $\leftrightarrow$ database workings

### CODE REVIEW

File: crawler/crawler.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+class Crawler:
+
+    def __init__(self):
+        self.url_base = "https://campus.tum.de/tumonline/pl/ui/$ctx"
+        self.german_p_session_id = ''.join(random.choices(string.ascii_uppercase + string.digits, k=64))
+        self.english_p_session_id = ''.join(random.choices(string.ascii_uppercase + string.digits, k=64))
+        self._switch_to_english(self.english_p_session_id)
+
+    def _get_p_session_id(self, english: bool):
```
Comment (severinsch) [inline crawler/crawler.py:22]: return type hint missing

File: crawler/main.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        self.database = database
+        self.module_id = module_id
+
+    def run(self):
+        for mapping in self.crawler.module_degree_mappings(self.module_id, english=True):
+            # TODO: add mapping to database
+            pass
+
+
+def run_crawler(self, max_workers=1):
```
Comment (severinsch) [inline crawler/main.py]: What is self here? The function is not contained in a class?

**Review by severinsch — COMMENTED** [review]
All functions are missing return type annotations.
Otherwise LGTM, provided it works

File: crawler/main.py (hunk 2 above)
Comment (c-rotte) [reply crawler/main.py]: > What is self here? The function is not contained in a class?

💀

**Comment by mkienitz** [comment]
Code for fetching data and parsing data should be separated whenever possible. In the current form this is not as composable as it could be
**λ**
Otherwise LGTM, provided it works

**Comment by c-rotte** [comment]
Exemplary calls:
```
degrees():
[{
  "text": "16 417 Production and Logistics (20131, Master's program, discontinued) [20131] Master's Program Production and Logistics (Curriculum version)",
  "id": "4343"
}, ...]


modules():
['1005861', 'Generalized Model Solutions for Physical Systems, Modeled by PDE's and Their Linear Stability']

module_degree_mappings()
[{
  "full_name": "1630 16 010 Mathematics ( Master's program)",
  "curriculum_version": "20191",
  "ects": 6.0,
  "weighting_factor": 1.0,
  "valid_from": "2020S",
  "valid_to": None,
  "degree_id": "4852"
}, ...]
```

**Comment by c-rotte** [comment]
image
First database scheme (doesn't include all of the additional data)

**Comment by c-rotte** [comment]
image
updated scheme

**Comment by c-rotte** [comment]
TODO: omit `full_text_*` in `module_degree_mappings`

**Comment by c-rotte** [comment]
First runthrough:
- 344 degrees
- 17292 modules

**Comment by c-rotte** [comment]
2 notes:
- Crawler seems to be broken: no mappings are added
- sometimes, modules with preexisting ids are added to the database, need to check on that

**Comment by lunar-seal** [comment]
@severinsch
use an url lib, e.g. [furl](https://github.com/gruns/furl) instead of manually assembling urls

**Comment by c-rotte** [comment]
Crawler seems to be working:
```
{
  "number_of_degrees": 1203,
  "number_of_modules": 17280,
  "number_of_mappings": 120967
}
```

**Review by lunar-seal — CHANGES_REQUESTED** [review]
Seems to work
Crawler could be improved a bit with 
- asyncio
- using furl
- use the cookies named arg of requests instead of setting the cookie header manually
- use sessions to prevent e.g. setting cookies each time manually

**Comment by c-rotte** [comment]
Added module numbers + fixed some bugs

**Comment by c-rotte** [comment]
TODO:
- [ ]  api endpoints
- [ ] refactoring

**Comment by c-rotte** [comment]
Last run took 7h 28m

**Comment by c-rotte** [comment]
```
$ select count(*) from degree;
+----------+
| count(*) |
+----------+
|     1203 |
+----------+

$ select count(*) from module;
+----------+
| count(*) |
+----------+
|    17280 |
+----------+

$ select count(*) from mapping;
+----------+
| count(*) |
+----------+
|   162550 |
+----------+
```

**Comment by c-rotte** [comment]
Seems to grab all versions:
```
$ select * from module where module_nr='IN2209';
+-----------+----------------+----------------+-----------+
| module_id | module_name_en | module_name_de | module_nr |
+-----------+----------------+----------------+-----------+
|    487469 | IT Security    | IT Sicherheit  | IN2209    |
|   1697995 | IT Security    | IT Sicherheit  | IN2209    |
+-----------+----------------+----------------+-----------+
```

**Comment by c-rotte** [comment]
[dump.zip](https://github.com/c-rotte/tum-ects/files/11365469/dump.zip)

**Pull request merged**